### PR TITLE
Import Remote Lambdas for supported interpreted languages

### PR DESCRIPTION
**Pull request by bryceitoc9**
## Description
Import remote NodeJS and Python Lambdas into the workspace folder. Runtimes must be current (e.g. no Node.js 0.10).
* Not supported for compiled languages (the downloaded artifact is compiled and worthless for authoring) or unsupported languages (e.g. Ruby)
* Saved to a workspace folder (or adds a folder to the workspace)
* Does not include any SAM-like trappings
* Creates a deduped entry in `launch.json` for immediate run/debug

## Motivation and Context
We have seen multiple requests for downloading remote Lambdas. This isn't generally considered best practice but is an oft-used feature in AWS Cloud9 and the Lambda console and will be followed by non-SAM direct Lambda deployment functionality.

## Related Issue(s)

https://github.com/aws/aws-toolkit-vscode/issues/105

## Testing

Manual tests, some automated tests (still trying to add more although unsure how much of this is unit-testable)

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the changelog using the script `npm run newChange`

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by justinmk3** [comment]
i would suggest putting the "Import..." action in the same group on the menu. 

- It's easy to miss in its current place. 
- It is an action similar to "Invoke". The current grouping seems far more arbitrary

here's what I would expect:

Screen Shot 2020-08-12 at 9 53 36 AM

**Comment by bryceitoc9** [comment]
Can do re: screenshot.

> Other comments
> 
>     * progress bar seems to work, nice!
> 

It "works" in that it's a 10% increment on starting download, a 70% increment on finishing download, a 10% increment on unzipping, and then it closes after any file-finding/launch config editing. There's no actual progress indicator provided by the SDK 😛  ; personally I would have preferred a "busy" icon but this does well enough.

>     * For the case where local files are _not_ being overwritten, we don't need a confirmation dialog ("Are you sure you want to import")? The cost of an accidental click is low, because the user _already_ had to choose a folder, which is simultaneously an opportunity to cancel the operation.
> 

Can do, this was a holdover from the existing branch.

File: src/shared/utilities/workspaceUtils.ts [hunk 1]
```diff
@@ -17,6 +17,15 @@ export function tryGetAbsolutePath(folder: vscode.WorkspaceFolder | undefined, r
     return path.resolve(folder?.uri ? folder.uri.fsPath + '/' : '', relPath)
 }
 
+export async function addWorkspaceFolder(folder: { uri: vscode.Uri; name?: string }): Promise<void> {
+    // No-op if the folder is already in the workspace.
+    if (vscode.workspace.getWorkspaceFolder(folder.uri)) {
+        return
+    }
+
+    await addFolderToWorkspace(folder)
```
Comment (justinmk3) [inline src/shared/utilities/workspaceUtils.ts:26]: we have `addFolderToWorkspace` and `addWorkspaceFolder`. Two similar functions with no real distinction except the no-op behavior. We can eliminate the extra `addWorkspaceFolder` function, and just add an optional parameter to `addFolderToWorkspace`:

```typescript
/** ...
 * @param skipExisting  Do nothing if `folder` is already a workspace
*/
export async function addFolderToWorkspace(folder: { uri: vscode.Uri; name?: string }, skipExisting?: boolean): ... {
    // No-op if the folder is already in the workspace.
    if (!!skipExisting && vscode.workspace.getWorkspaceFolder(folder.uri)) {
        return
    }
}
```

I know this was old code, but... since we're already moving the function, it's a good time to just eliminate it.

File: src/lambda/models/samLambdaRuntime.ts [hunk 2]
```diff
@@ -30,6 +30,7 @@ const DEFAULT_RUNTIMES = Map<RuntimeFamily, Runtime>([
 ])
 
 export const samLambdaRuntimes: Set<Runtime> = Set.union([nodeJsRuntimes, pythonRuntimes, dotNetRuntimes])
+export const importableSamLambdaRuntimes: Set<Runtime> = Set.union([nodeJsRuntimes, pythonRuntimes])
```
Comment (justinmk3) [inline src/lambda/models/samLambdaRuntime.ts:33]: we have `samLambdaCreatableRuntimes` already, so it is more consistent to name this `samLambdaImportableRuntimes`. It also helps discoverability (and as a visual aid): these are similar concepts, all having the same prefix (`samLambdaX`)--not to mention, the same casing.

File: src/shared/clients/lambdaClient.ts [hunk 3]
```diff
@@ -14,4 +14,6 @@ export interface LambdaClient {
     invoke(name: string, payload?: _Blob): Promise<Lambda.InvocationResponse>
 
     listFunctions(): AsyncIterableIterator<Lambda.FunctionConfiguration>
+
+    getFunction(name: string): Promise<Lambda.GetFunctionResponse>
```
Comment (justinmk3) [inline src/shared/clients/lambdaClient.ts:18]: not sure if we don't document these because they closely match the SDK, but this one is a bit non-obvious to me, so I would add a brief docstring.

```
/** Downloads the Lambda function source code as a zip. */

File: .changes/next-release/Feature-c56b7fa8-7102-4006-afd1-5458cb70c0ef.json [hunk 4]
```diff
@@ -0,0 +1,4 @@
+{
+    "type": "Feature",
+    "description": "Users can import supported NodeJS and Python Lambda functions"
```
Comment (awschristou) [inline .changes/next-release/Feature-c56b7fa8-7102-4006-afd1-5458cb70c0ef.json:3]:
```suggestion
    "description": "Users can import supported NodeJS and Python Lambda functions from their AWS accounts into their local workspace"
```

File: src/shared/clients/lambdaClient.ts (hunk 3 above)
Comment (bryceitoc9) [reply src/shared/clients/lambdaClient.ts:18]: FYI this doesn't do that, this just wraps the getFunction call, which returns metadata + a link to the function's zip file.

File: package.json [hunk 5]
```diff
@@ -1095,6 +1110,7 @@
         "cross-spawn": "^6.0.5",
         "del": "^3.0.0",
         "fs-extra": "^8.1.0",
+        "got": "^11.5.1",
```
Comment (awschristou) [inline package.json:1113]: Looks like we currently use the `request` package. Why are we introducing another package with similar responsibilities?

File: src/awsexplorer/activation.ts [hunk 6]
```diff
@@ -162,6 +163,13 @@ async function registerAwsExplorerCommands(
         )
     )
 
+    context.subscriptions.push(
+        vscode.commands.registerCommand(
+            'aws.importLambda',
+            async (node: LambdaFunctionNode) => await importLambdaCommand(node)
```
Comment (awschristou) [inline src/awsexplorer/activation.ts:169]: Is there telemetry for this command/operation?

File: src/lambda/activation.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import { AwsExplorer } from '../awsexplorer/awsExplorer'
+import { ext } from '../shared/extensionGlobals'
+import { deleteLambda } from './commands/deleteLambda'
+import { invokeLambda } from './commands/invokeLambda'
+import { LambdaFunctionNode } from './explorer/lambdaFunctionNode'
+
+/**
+ * Activates Lambda components.
+ */
+export async function activate(extensionContext: vscode.ExtensionContext, awsExplorer: AwsExplorer): Promise<void> {
```
Comment (awschristou) [inline src/lambda/activation.ts:16]: 1. where is this activated from? I don't see a change to extension.ts
2. why is importLambda registered in src/awsexplorer/activation.ts but other lambda commands are registered here?
3. are the delete and invoke command registrations removed from elsewhere?

File: src/awsexplorer/activation.ts (hunk 6 above)
Comment (bryceitoc9) [reply src/awsexplorer/activation.ts:169]: Not right now, will add.

File: src/shared/codelens/codeLensUtils.ts [hunk 8]
```diff
@@ -66,7 +67,11 @@ export async function makeCodeLenses({
                 rootUri: handler.manifestUri,
                 runtimeFamily,
             }
-            if (!existingConfigs.has(path.join(path.dirname(baseParams.rootUri.fsPath), baseParams.handlerName))) {
+            if (
+                !existingConfigs.has(
+                    pathutils.normalize(path.join(path.dirname(baseParams.rootUri.fsPath), baseParams.handlerName))
```
Comment (awschristou) [inline src/shared/codelens/codeLensUtils.ts:72]: nit: Recommend moving this change to separate PR if not related to this one.

File: src/shared/clients/lambdaClient.ts (hunk 3 above)
Comment (justinmk3) [reply src/shared/clients/lambdaClient.ts:18]: docstring will clear that up then :)

File: src/lambda/commands/importLambda.ts [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+import { localize } from '../../shared/utilities/vsCodeUtils'
+import { Window } from '../../shared/vscode/window'
+import { LambdaFunctionNode } from '../explorer/lambdaFunctionNode'
+import { getFamily, RuntimeFamily } from '../models/samLambdaRuntime'
+import { showConfirmationMessage } from '../../s3/util/messages'
+import { addWorkspaceFolder } from '../../shared/utilities/workspaceUtils'
+
+const pipeline = promisify(Stream.pipeline)
+
+export async function importLambdaCommand(functionNode: LambdaFunctionNode, window = Window.vscode()) {
```
Comment (awschristou) [inline src/lambda/commands/importLambda.ts:31]: 1. function accepts `window` but still makes direct vscode calls
2. this signature makes me believe the function is set up for testing but has no test coverage

File: src/lambda/commands/importLambda.ts [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+            openLabel: 'Select',
+            canSelectFiles: false,
+            canSelectFolders: true,
+        })
+        if (!selectedDirectory) {
+            getLogger().info('ImportLambda cancelled')
+            return
+        }
+        selectedUri = selectedDirectory[0]
+        addWorkspaceFolder({ uri: selectedUri })
```
Comment (awschristou) [inline src/lambda/commands/importLambda.ts:61]: Unless things have changed, if user doesn't have a workspace opened, adding a workspace folder will restart vs code and the extension, which likely kills this function. If that is the case, consider quitting with a message at the start of the command, asking user to create/open a workspace first.

File: src/lambda/commands/importLambda.ts [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+    const importLocationName = vscode.workspace.asRelativePath(importLocation, true)
+    const workspaceFolder = isWorkspaceFolder
+        ? labelToWorkspace.get(selectedLocation)!
+        : vscode.workspace.getWorkspaceFolder(vscode.Uri.file(importLocation))!
+
+    const directoryExists = await fs.pathExists(importLocation)
+
+    const overwriteWarning = localize(
+        'AWS.lambda.import.overwriteWarning',
+        '\nA directory named {0} already exists! Importing will overwrite any existing files!\n',
```
Comment (awschristou) [inline src/lambda/commands/importLambda.ts:76]: Message could be made more concise and less exclamatory:

`{0} already exists. Existing files will be overwritten.`

File: src/lambda/commands/importLambda.ts [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+        functionName
+    )
+    const isConfirmed = await showConfirmationMessage(
+        {
+            prompt: localize(
+                'AWS.lambda.import.prompt',
+                'This will import {0} into {1}.\n{2}\nAre you sure you want to import the function?',
```
Comment (awschristou) [inline src/lambda/commands/importLambda.ts:83]: Message could be made more concise:

`About to import {0} into {1}`
`Proceed with import?`

File: src/lambda/commands/importLambda.ts [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+        : vscode.workspace.getWorkspaceFolder(vscode.Uri.file(importLocation))!
+
+    const directoryExists = await fs.pathExists(importLocation)
+
+    const overwriteWarning = localize(
+        'AWS.lambda.import.overwriteWarning',
+        '\nA directory named {0} already exists! Importing will overwrite any existing files!\n',
+        functionName
+    )
+    const isConfirmed = await showConfirmationMessage(
```
Comment (awschristou) [inline src/lambda/commands/importLambda.ts:79]: Please confirm that this message is modal. Waiting for non-modal messages that can be queued and acted on later can cause confusion.

File: src/lambda/commands/importLambda.ts [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+            ),
+        },
+        async progress => {
+            const lambdaLocation = path.join(importLocation, getLambdaFileNameFromHandler(functionNode.configuration))
+            try {
+                await downloadAndUnzipLambda(progress, functionNode, importLocation)
+                await openLambdaFile(lambdaLocation)
+                await addLaunchConfigEntry(lambdaLocation, functionNode, workspaceFolder)
+            } catch (e) {
+                // swallow error; all functions handle errors themselves
```
Comment (awschristou) [inline src/lambda/commands/importLambda.ts:117]: It may be worth making less assumptions and at least logging something.

File: src/lambda/commands/importLambda.ts [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+        functionName
+    )
+    const isConfirmed = await showConfirmationMessage(
+        {
+            prompt: localize(
+                'AWS.lambda.import.prompt',
+                'This will import {0} into {1}.\n{2}\nAre you sure you want to import the function?',
+                functionName,
+                importLocationName,
+                directoryExists ? overwriteWarning : ''
```
Comment (awschristou) [inline src/lambda/commands/importLambda.ts:86]: The message box should be a warning when an overwrite is detected

File: src/lambda/commands/importLambda.ts (hunk 13 above)
Comment (bryceitoc9) [reply src/lambda/commands/importLambda.ts:79]: Confirming this is a modal. Can revisit this + wording.

File: src/lambda/commands/importLambda.ts (hunk 11 above)
Comment (justinmk3) [reply src/lambda/commands/importLambda.ts:76]: could also put the variable part at the end, which helps the reader (purpose of the message is always in the same visual location, instead of varying based on filename):

    Directory already exists and will be overwritten: {0}

File: package.json (hunk 5 above)
Comment (bryceitoc9) [reply package.json:1113]: Looks like `request` is deprecated: https://www.npmjs.com/package/got ; I'll see if it's worth moving off of that to this.

Comment (awschristou) [reply package.json:1113]: What if we prioritize a migration from `request` separate from this change, and use request for the time being? It looks like it was deprecated six months ago, but I don't imagine a migration is necessary for this PR.

File: src/lambda/commands/importLambda.ts [hunk 16]
```diff
@@ -25,49 +25,30 @@ import { LambdaFunctionNode } from '../explorer/lambdaFunctionNode'
...
-        placeHolder: localize('AWS.lambda.import.prompt.placeholder', 'Select the import location'),
-    })
-
-    if (!selectedLocation) {
-        getLogger().info('ImportLambda cancelled')
+    if (workspaceFolders.length === 0) {
+        window.showErrorMessage(
+            localize(
+                'AWS.lambda.import.noWorkspaceFolders',
+                'Please add a workspace folder before importing a Lambda function.'
```
Comment (justinmk3) [inline src/lambda/commands/importLambda.ts:38]: "Please" is never needed in messages, better to draw attention to the actual content of the message. User won't be offended :)

    'Select a workspace before importing a Lambda function."

File: src/shared/wizards/multiStepWizard.ts [hunk 17]
```diff
@@ -90,3 +95,54 @@ export class BrowseFolderQuickPickItem implements FolderQuickPickItem {
...
+    helpButton?: { button: vscode.QuickInputButton; url: string }
+): Promise<vscode.Uri | undefined> {
+    const items: FolderQuickPickItem[] = (context.workspaceFolders || [])
+        .map<FolderQuickPickItem>(f => new WorkspaceFolderQuickPickItem(f))
+        .concat([
+            new BrowseFolderQuickPickItem(
+                context,
+                localize(
+                    'AWS.samcli.initWizard.location.prompt',
+                    'The folder you select will be added to your VS Code workspace.'
```
Comment (justinmk3) [inline src/shared/wizards/multiStepWizard.ts:110]: don't need to mention the editor name, just "your workspace" or "the workspace" is enough.

File: src/eventSchemas/wizards/schemaCodeDownloadWizard.ts [hunk 18]
```diff
@@ -122,54 +115,7 @@ export class DefaultSchemaCodeDownloadWizardContext extends WizardContext implem
...
-
-        if (pickerResponse instanceof BrowseFolderQuickPickItem) {
-            const browseFolderResult = await pickerResponse.getUri()
-
-            // If user cancels from Open Folder dialog, send them back to the folder picker.
-            return browseFolderResult ? browseFolderResult : this.promptUserForLocation()
-        }
-
-        return pickerResponse.getUri()
+        return promptUserForLocation(this, { button: this.helpButton, url: eventBridgeSchemasDocUrl })
```
Comment (justinmk3) [inline src/eventSchemas/wizards/schemaCodeDownloadWizard.ts:118]: nice code reduction!

File: src/shared/codelens/codeLensUtils.ts (hunk 8 above)
Comment (bryceitoc9) [reply src/shared/codelens/codeLensUtils.ts:72]: I forget what this was for, might have been for upload Lambdas (and I forgot to swap the branch). I'll give it a look.

File: src/lambda/commands/importLambda.ts (hunk 15 above)
Comment (awschristou) [reply src/lambda/commands/importLambda.ts:63]: Is warning appropriate when there is no overwrite? If I've gestured to do an import, I'd expect the prompt to be neutral (question/information). If I was presented with a warning or error, I'd get the impression there is something I should be concerned about.

Comment (justinmk3) [reply src/lambda/commands/importLambda.ts:63]: > If I've gestured to do an import, I'd expect the prompt to be neutral (question/information)

We don't need a prompt at all in that case, see https://github.com/aws/aws-toolkit-vscode/pull/1236#issuecomment-672993443

Comment (awschristou) [reply src/lambda/commands/importLambda.ts:63]: I agree with that assessment @justinmk3 ... easy to get tunnel vision looking at a piece of code in isolation 😄

File: package.nls.json [hunk 19]
```diff
@@ -186,15 +186,14 @@
     "AWS.lambda.debug.node.launchConfig.name": "Lambda: Debug {0} locally",
     "AWS.lambda.debug.node.attachConfig.name": "Lambda: Attach to {0} locally",
     "AWS.lambda.debug.node.invokeTask.label": "Lambda: Invoke {0} locally",
-    "AWS.lambda.import.prompt": "About to import {0} into {1}.\n{2}\nProceed with import?'",
+    "AWS.lambda.import.prompt": "About to import {0} into {1}.\nExisting directory will be overwritten: {0}\nProceed with import?",
```
Comment (justinmk3) [inline package.nls.json:189]:
```suggestion
    "AWS.lambda.import.prompt": "Importing {0} into: {1}\nExisting directory will be overwritten: {0}\nProceed with import?",
```

File: src/shared/wizards/multiStepWizard.ts [hunk 20]
```diff
@@ -98,36 +98,46 @@ export class BrowseFolderQuickPickItem implements FolderQuickPickItem {
...
                 context,
-                localize(
-                    'AWS.samcli.initWizard.location.select.folder.detail',
-                    'The selected folder will be added to the VS Code workspace.'
-                )
+                additionalParams?.overrideText?.detail
+                    ? additionalParams.overrideText.detail
+                    : localize(
+                          'AWS.wizard.location.select.folder.detail',
+                          'The selected folder will be added to the VS Code workspace.'
```
Comment (justinmk3) [inline src/shared/wizards/multiStepWizard.ts:115]: `getIdeName()` instead of "VS Code" ?
